### PR TITLE
Test: Added unit tests for Catalyst Separation Plugin (1) - c8s65gxjrsjgb3brm8im6h9c74c

### PR DESCRIPTION
**Pull request by Hanan-Nawaz**
**Description**

- Updated and added unit tests for Catalyst separation Plugin.

**Motivation / Background**

- Makes tests more readable, maintainable, and flexible.
- Ensures correct yearly cost insertion into DCF and proper calculations for edge cases.

**Type of Change**

- [x] Test addition or update

**How Has This Been Tested?**

- All tests in test_catalyst_separation.py pass with uv run pytest -v.

**Checklist**

- [x]  Tests added / updated and passing
- [x]  Comments added in critical sections

**Screenshot**
Capture1

### CODE REVIEW

File: src/tests/plugins/test_catalyst_separation_plugin.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    [
+        (0.0, 50.0, 5.0, 0),  # zero water
+        (10000.0, 0.0, 5.0, 0),  # zero filtration cost
+        (0.0, 0.0, 5.0, 0),  # both zero
+        (1000.0, 50.0, 2.0, 25.0),  # real cases
+    ],
+)
+
+
+def test_catalyst_separation_plugin(water_volume_liters, filtration_cost_per_m3, catalyst_lifetime_years, expected_result):
```
Comment (jschneidewind) [inline src/tests/plugins/test_catalyst_separation_plugin.py]: Where is expected_result defined? At the moment, I am missing the expected result set from the test file

**Review by jschneidewind — CHANGES_REQUESTED** [review]

File: src/tests/plugins/test_catalyst_separation_plugin.py (hunk 1 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/test_catalyst_separation_plugin.py]: We have to hard code the "expected result" in the parameters, as we discussed we are getting the result from the master branch, so we just add the Result in the 3 index like here  (1000.0, 50.0, 2.0, 25.0) 25 is result,  we added and rest pytest handled it. In the case of catalyst_separation_plugin, the calculations are easy, so i added the edge cases, for others it will be just one case that will be the case from master branch.

@pytest.mark.parametrize(
    "water_volume_liters, filtration_cost_per_m3, catalyst_lifetime_years, expected_result",
    [
        (0.0, 50.0, 5.0, 0),  # zero water
        (10000.0, 0.0, 5.0, 0),  # zero filtration cost
        (0.0, 0.0, 5.0, 0),  # both zero
        (1000.0, 50.0, 2.0, 25.0),  # real cases
    ],
)

Comment (jschneidewind) [reply src/tests/plugins/test_catalyst_separation_plugin.py]: Makes sense, I missed that the last entry of the set is the expected result!

File: src/tests/plugins/catalyst_separation_plugin_test.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import pytest
```
Comment (gulzerr) [inline src/tests/plugins/catalyst_separation_plugin_test.py:1]: Please, change the file name to `catalyst_separation_plugin_test.py`. this is the standard practice

File: src/tests/plugins/test_catalyst_separation_plugin.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        }
+
+
+@pytest.mark.parametrize(
+    "water_volume_liters, filtration_cost_per_m3, catalyst_lifetime_years, expected_result",
+    [
+        (0.0, 50.0, 5.0, 0),  # zero water
+        (10000.0, 0.0, 5.0, 0),  # zero filtration cost
+        (0.0, 0.0, 5.0, 0),  # both zero
+        (1000.0, 50.0, 2.0, 25.0),  # real cases
```
Comment (gulzerr) [inline src/tests/plugins/test_catalyst_separation_plugin.py]: 1. If I am not wrong, `catalyst_lifetime_years` is the divider. What if the user input 0 (zero)? Is it handled in the plugin?
2. What if user input is a string?

Were these edge cases covered? If these values are not given by users, it's fine.

File: src/tests/plugins/catalyst_separation_plugin_test.py (hunk 2 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/catalyst_separation_plugin_test.py:1]: Ok, I will change the name. But as far as I know, both are fine. 

Reference 1: https://pytest-with-eric.com/pytest-best-practices/pytest-organize-tests/
Reference 2: https://docs.pytest.org/en/stable/explanation/goodpractices.html

File: src/tests/plugins/test_catalyst_separation_plugin.py (hunk 3 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/test_catalyst_separation_plugin.py]: 1. For the first case, No, it didn't handled by the plugin. we have the divider 
     1./dcf.inp['Catalyst']['Lifetime (years)']['Value']
     if the dcf.inp['Catalyst']['Lifetime (years)']['Value'] is zero, it will for sure give an error (Zero Division Error).

2. It is also not handled by the plugin. I just checked it.

Comment (Hanan-Nawaz) [reply src/tests/plugins/test_catalyst_separation_plugin.py]: our main purpose is not to check the correctness of plugin, but to see when we change the code in plugin, whether it performs the same way:

"Therefore, the tests will by definition all pass for the Master branch version. The idea with these plugin tests is that when we make changes to the plugins, they give us feedback if we did anything to the plugins that breaks their behaviour compared to the Master branch behaviour. The idea at this point is not to check the correctness of every single function/computation by computing their expected value by hand (this would go too far at this stage)" -Jacob

File: src/tests/plugins/test_catalyst_separation_plugin.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            "Water Volume": {"Volume (liters)": {"Value": water_volume_liters}},
+            "Catalyst": {"Lifetime (years)": {"Value": catalyst_lifetime_years}},
+            "Catalyst Separation": {
+                "Filtration cost ($/m3)": {"Value": filtration_cost_per_m3}
+            },
+        }
+
+
+@pytest.mark.parametrize(
+    "water_volume_liters, filtration_cost_per_m3, catalyst_lifetime_years, expected_result",
```
Comment (jschneidewind) [inline src/tests/plugins/test_catalyst_separation_plugin.py]: I think it is better if we define the inputs and expected result as a dictionary, of the form:

```suggestion
    input_expected_output = {'water_volume_liters': 0,
                                                 'filtration_cost_per_m3': 50.0,
                                                 'catalyst_lifetime_years': 5.0,
                                                 'expected_result': 0}
```
This we can then pass to test_catalyst_separation_plugin via **input_expected_output (**kwargs passing). I think the dictionary structure is needed for more complicated plugins, where the inputs might be arrays or themselves dicts. For those complicated inputs the set structure will quickly get confusing (yes that means we have to always declare a dict for each test case, but I think it is better to be more rather than less explicit for the tests)

**Review by jschneidewind — CHANGES_REQUESTED** [review]

File: src/tests/plugins/test_catalyst_separation_plugin.py (hunk 4 above)
Comment (Hanan-Nawaz) [reply src/tests/plugins/test_catalyst_separation_plugin.py]: @pytest.mark.parametrize(
    "case",
    [
        {
            "input": {
                "water_volume_liters": 0.0,
                "filtration_cost_per_m3": 50.0,
                "catalyst_lifetime_years": 5.0,
            },
            "expected": 0,
        },
        {
            "input": {
                "water_volume_liters": 0.0,
                "filtration_cost_per_m3": 0.0,
                "catalyst_lifetime_years": 5.0,
            },
            "expected": 0,
        },
        {
            "input": {
                "water_volume_liters": 10000.0,
                "filtration_cost_per_m3": 0.0,
                "catalyst_lifetime_years": 5.0,
            },
            "expected": 0,
        },
        {
            "input": {
                "water_volume_liters": 1000.0,
                "filtration_cost_per_m3": 50.0,
                "catalyst_lifetime_years": 2.0,
            },
            "expected": 25.0,
        },
    ],
)

**Review by jschneidewind — COMMENTED** [review]
Please replace the uv.lock file by the one from master branch and then run "uv sync" to refresh the uv.lock file, this should resolve the merge conflict

**Review by jschneidewind — APPROVED** [review]
Looks good, thanks!
